### PR TITLE
Upgrading ctr to 1.3.2 and unmounting disk on container shutdown

### PR DESCRIPTION
**Pull request by adarsh-zededa**
Signed-off-by: adarsh-zededa <adarsh@zededa.com>

### CODE REVIEW

**Comment by deitch** [comment]
This is really important. Apparently we had quite an old version of containerd running (1.1.1), which had some bugs, notably the lingering snapshot that adarsh and I spent too many days tracking down.

File: images/rootfs-acrn.yml.in [hunk 1]
```diff
@@ -3,8 +3,8 @@ kernel:
   cmdline: "rootdelay=3"
 init:
   - linuxkit/init:v0.5
-  - linuxkit/runc:v0.5
-  - linuxkit/containerd:v0.5
+  - linuxkit/runc:69b4a35eaa22eba4990ee52cccc8f48f6c08ed03
+  - linuxkit/containerd:09553963ed9da626c25cf8acdf6d62ec37645412
```
Comment (rvs) [inline images/rootfs-acrn.yml.in]: Can we use the latest release of linuxkit here or do we have to latch onto this?

**Comment by rvs** [comment]
> This is really important. Apparently we had quite an old version of containerd running (1.1.1), which had some bugs, notably the lingering snapshot that adarsh and I spent too many days tracking down.

Speaking of which @deitch -- what's the status of that /dev disappearing issue you were tracking? It actually will become a problem in a few weeks from now when we plan to start using top-level container management for system containers much more dynamically.

**Comment by deitch** [comment]
> Speaking of which @deitch -- what's the status of that /dev disappearing issue you were tracking? It actually will become a problem in a few weeks from now when we plan to start using top-level container management for system containers much more dynamically.

Oh shoot. Totally forgot about it @rvs . We first noticed it ~7 weeks ago, figured out the problem ~5.5-6 weeks ago and opened an issue on containerd. Sebastian and Phil figured out the root of the bad code, but there have been no updates since. I will bump it there, but let us also try to think of a workaround.

File: images/rootfs-acrn.yml.in (hunk 1 above)
Comment (adarsh-zededa) [reply images/rootfs-acrn.yml.in]: I tried with the latest version on linuxkit (v0.7) and faced the same issue. But with the latest tags as per https://hub.docker.com/r/linuxkit/runc/tags and https://hub.docker.com/r/linuxkit/containerd/tags snapshot issue was resolved.

**Comment by deitch** [comment]
> I tried with the latest version on linuxkit (v0.7) and faced the same issue

Can you check in the linuxit pkg source [here](https://github.com/linuxkit/linuxkit/tree/master/pkg/containerd) and the history, see what versions of containerd it is that resolves it? That should help us track down the actual issue that resolved it. I can push to cut a new versioned release of formal linuxkit for that then.

File: images/rootfs-acrn.yml.in (hunk 1 above)
Comment (rvs) [reply images/rootfs-acrn.yml.in]: Oh, ok

**Comment by rvs** [comment]
OK, I'll merge this now -- but hopefully @adarsh-zededa you can answer @deitch question asynchronously, ok?

**Comment by adarsh-zededa** [comment]
> > I tried with the latest version on linuxkit (v0.7) and faced the same issue
> 
> Can you check in the linuxit pkg source [here](https://github.com/linuxkit/linuxkit/tree/master/pkg/containerd) and the history, see what versions of containerd it is that resolves it? That should help us track down the actual issue that resolved it. I can push to cut a new versioned release of formal linuxkit for that then.

About ctrd snapshot clean-up issue:
Q1) Why were we able to delete snapshot from CLI but not from eve (API) earlier?

We have to specifically mention an option to delete snapshot while deleting the container through API. I had missed that earlier. That is the reason ctr deleted the snapshot (because ctr passes that option when deleting the container) but eve didn’t. Tested it by rolling back containerd to v1.1.1 and added that snapshot delete option while deleting container and it worked.


Q2) Which version of containerd solves that?

containerd v1.3.0: https://github.com/containerd/containerd/releases/tag/v1.3.0


Q3) Which linuxkit/containerd has containerd v1.3.0?

linuxkit/containerd version: https://hub.docker.com/layers/linuxkit/containerd/edb204039976291ce0b60547193f57696981ab37/images/sha256-285c5bee8ce32f0a9f8e004ef2db31aff307456908dd28703402e0fca0f1ded3?context=explore


Q4) How was this resolved in containerd v1.3.0?

There were a LOT of changes to improve the overall GC in this release. Because of too many changes, It’s hard to narrow down which commit fixed the issue but going through all snapshot and GC related commits I guess this particular commit solved the issue: https://github.com/containerd/containerd/commit/75f183887a99de2fd29d3144e5f9a1e09a069a67#diff-491f07c28eb851b720f11ce0d3dbb4f8 . When GC is scheduled (in one way) it goes over the labels and decides what and what not to delete and prior to this commit, few labels were not passed over on creation of snapshot which may have caused gc to skip the snapshot.

**Pull request merged**